File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;

void dfs(vector<int>& numbers, int target, int curSum, int curIdx)
{
    if (curIdx == numbers.size())
    {
        if (target == curSum)
            ++answer;
        return;
    }
    
    dfs(numbers, target, curSum + numbers[curIdx], curIdx + 1);
    dfs(numbers, target, curSum - numbers[curIdx], curIdx + 1);
}
int solution(vector<int> numbers, int target) {
    
    dfs(numbers, target, 0, 0);
    
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    public static int answer = 0;
    
    public int solution(int[] numbers, int target) {
        dfs(0, 0, numbers, target);    
        return answer;
    }
    
    public void dfs(int idx, int num, int[] numbers, int target) {
        if (idx == numbers.length) {
            if (num == target) {
                answer++;
            }
            return;
        }

        
        dfs(idx + 1, num + numbers[idx], numbers, target);
        dfs(idx + 1, num - numbers[idx], numbers, target);
    }
}